lib/puppet/catalog-diff/compilecatalog.rb

```diff
@@ -122,6 +125,43 @@ def render_pson(catalog)
       pson
     end
 
+    def clean_sensitive_parameters!(catalog)
+      catalog['resources'].map! do |resource|
+        if resource.key? 'sensitive_parameters'
+          resource['sensitive_parameters'].each do |p|
+            hash = Digest::SHA256.hexdigest Marshal.dump(resource['parameters'][p])
+            resource['parameters'][p] = "Sensitive [hash #{hash}]"
+          end
+          resource.delete('sensitive_parameters')
+          resource
+        else
+          resource
+        end
+      end
+    end
+
+    def clean_nested_sensitive_parameters!(catalog)
+      # Resources can also contain sensitive data nested deep in hashes/arrays
+      catalog['resources'].each do |resource|
+        redact_sensitive(resource['parameters']) if resource.key? 'parameters'
+      end
+    end
+
+    def redact_sensitive(data)
+      if data.is_a?(Hash) && data.key?('__ptype')
+        data[:catalog_diff_hash] = Digest::SHA256.hexdigest Marshal.dump(data['__pvalue'])
```

Comment:
Ideally, I think I would have liked to replace this hash with a `"Sensitive [hash #{hash}]"` style string (instead of replacing the `__ptype` and `__pvalue` keys.  I couldn't figure out how to do this in place though.